### PR TITLE
retry current request on "Communications link failure" in MDB

### PR DESCRIPTION
**Pull request by anandjakhaniya**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/batchProcess/OwnerRoleBindingRepositoryUtils.java [hunk 1]
```diff
@@ -124,6 +125,12 @@ private static void migrateRepositories() {
 
         transaction.commit();
         lowerBound += pagesize;
+      } catch (Exception ex) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/batchProcess/OwnerRoleBindingRepositoryUtils.java:128]: not related directly to the PR but can you please add comments on the top of the files in batchProcess/* on what they are doing?

Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/batchProcess/OwnerRoleBindingRepositoryUtils.java:128]: Sure. it makes sense.

**Review by ravishetye — COMMENTED** [review]
we spoke about the wrapper around session operators, did that not work?

**Comment by anandjakhaniya** [comment]
No, it does not work because each and every `session.*` operations need to retry with new sessions after the old session invalidates so for `session.update`, `session.save` etc. need to open transaction after acquiring new connection and it would create multiple transactions in many endpoints DAO.

File: backend/src/main/java/ai/verta/modeldb/utils/ModelDBUtils.java [hunk 2]
```diff
@@ -480,6 +482,28 @@ public static Throwable findRootCause(Throwable throwable) {
     responseObserver.onError(statusRuntimeException);
   }
 
+  public static boolean needToRetry(Exception ex) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/utils/ModelDBUtils.java:484]: we need to also cover retrying on Unavailable for UAC right?

Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/utils/ModelDBUtils.java:484]: `Unavailable for UAC` is a separate thing, we will apply the same logic in AuthService & RoleService for the same.

**Comment by ravishetye** [comment]
not all of mdb/uac APIs will be safe to be retried.  This would have open gate for other bugs which we would be required to fix. Is there anything better we could do in that area?

cc @conradoverta

File: backend/src/main/java/ai/verta/modeldb/utils/ModelDBUtils.java [hunk 3]
```diff
@@ -480,6 +481,32 @@ public static Throwable findRootCause(Throwable throwable) {
...
+    Throwable communicationsException = findCommunicationsFailedCause(ex);
+    if ((communicationsException.getCause() instanceof CommunicationsException)
+        || (communicationsException.getCause() instanceof SocketException)) {
+      LOGGER.warn(communicationsException.getMessage());
+      if (ModelDBHibernateUtil.ping()) {
+        ModelDBHibernateUtil.resetSessionFactory();
+      }
```
Comment (anandjakhaniya) [inline backend/src/main/java/ai/verta/modeldb/utils/ModelDBUtils.java]: @ravishetye added this new check to reset the session factory

**Review by ravishetye — APPROVED** [review]
LGTM , but I will test it more,before merging it.

**Review by evgeniilezhnin — APPROVED** [review]

**Comment by ravishetye** [comment]
Screen Shot 2020-06-02 at 8 35 48 PM

**Comment by ravishetye** [comment]
@ad-47 once you fix the NPE , I can merge it.
Good Job.

**Pull request merged**